File: services/r2Uploader.js
/**
 * Web2Media Service — Cloudflare R2 Upload Service
 */

const fs = require('fs');
const path = require('path');
const { S3Client } = require('@aws-sdk/client-s3');
const { Upload } = require('@aws-sdk/lib-storage');
const { SERVER_CONFIG } = require('../config');

function normalizeEndpoint(r2Config) {
  if (r2Config.endpoint) {
    return r2Config.endpoint.replace(/\/+$/, '');
  }

  if (r2Config.accountId) {
    return `https://${r2Config.accountId}.r2.cloudflarestorage.com`;
  }

  return null;
}

function normalizeKey(key) {
  return key
    .replace(/\\/g, '/')
    .replace(/^\/+/, '')
    .replace(/\/{2,}/g, '/');
}

function buildR2Key(r2Config, filename) {
  if (r2Config.key) {
    return normalizeKey(r2Config.key);
  }

  const normalizedPrefix = r2Config.keyPrefix ? normalizeKey(r2Config.keyPrefix) : '';
  const prefix = normalizedPrefix ? `${normalizedPrefix.replace(/\/?$/, '/')}` : '';

  return `${prefix}${filename}`;
}

function buildPublicUrl(r2Config, key) {
  if (!r2Config.publicBaseUrl) {
    return null;
  }

  const encodedKey = key.split('/').map(encodeURIComponent).join('/');
  return `${r2Config.publicBaseUrl.replace(/\/+$/, '')}/${encodedKey}`;
}

async function uploadFileToR2(filePath, r2Config, options = {}) {
  const endpoint = normalizeEndpoint(r2Config);
  if (!endpoint) {
    throw new Error('Thiếu endpoint hoặc accountId trong cấu hình R2 runtime_configs');
  }

  const filename = options.filename || path.basename(filePath);
  const key = buildR2Key(r2Config, filename);
  const contentType = options.contentType || 'application/octet-stream';
  const fileSize = fs.statSync(filePath).size;

  const client = new S3Client({
    region: r2Config.region || 'auto',
    endpoint,
    forcePathStyle: true,
    requestChecksumCalculation: 'WHEN_REQUIRED',
    responseChecksumValidation: 'WHEN_REQUIRED',
    credentials: {
      accessKeyId: r2Config.accessKeyId,
      secretAccessKey: r2Config.secretAccessKey,
    },
  });

  console.log(`[R2] Uploading ${filePath} -> ${r2Config.bucket}/${key}`);

  const partSize = r2Config.partSize || SERVER_CONFIG.r2UploadPartSize;
  const queueSize = r2Config.queueSize || SERVER_CONFIG.r2UploadQueueSize;

  const upload = new Upload({
    client,
    queueSize,
    partSize,
    leavePartsOnError: false,
    params: {
      Bucket: r2Config.bucket,
      Key: key,
      Body: fs.createReadStream(filePath),
      ContentLength: fileSize,
      ContentType: contentType,
    },
  });

  upload.on('httpUploadProgress', (progress) => {
    if (!progress.loaded || !progress.total) {
      return;
    }

    const percent = ((progress.loaded / progress.total) * 100).toFixed(1);
    console.log(`[R2] Upload progress: ${percent}% (${progress.loaded}/${progress.total} bytes)`);
  });

  const response = await upload.done();

  console.log(`[R2] Upload complete: ${r2Config.bucket}/${key}`);

  return {
    bucket: r2Config.bucket,
    key,
    size: fileSize,
    contentType,
    etag: response.ETag,
    url: buildPublicUrl(r2Config, key),
  };
}

module.exports = {
  uploadFileToR2,
};

File: app/services/r2_uploader.py
from __future__ import annotations

import mimetypes
from pathlib import Path
from typing import Any
from urllib.parse import quote

import boto3
from boto3.s3.transfer import TransferConfig
from botocore.config import Config

from app.core.config import SERVER_CONFIG


def normalize_endpoint(r2_config: dict[str, Any]) -> str | None:
    endpoint = r2_config.get("endpoint")
    if endpoint:
        return str(endpoint).rstrip("/")

    account_id = r2_config.get("accountId")
    if account_id:
        return f"https://{account_id}.r2.cloudflarestorage.com"

    return None


def normalize_key(key: str) -> str:
    normalized = key.replace("\\", "/").lstrip("/")
    while "//" in normalized:
        normalized = normalized.replace("//", "/")
    return normalized


def build_r2_key(r2_config: dict[str, Any], filename: str) -> str:
    if r2_config.get("key"):
        return normalize_key(str(r2_config["key"]))

    prefix = normalize_key(str(r2_config.get("keyPrefix") or ""))
    if prefix:
        prefix = prefix.rstrip("/") + "/"
    return f"{prefix}{filename}"


def build_public_url(r2_config: dict[str, Any], key: str) -> str | None:
    public_base_url = r2_config.get("publicBaseUrl")
    if not public_base_url:
        return None
    encoded_key = "/".join(quote(part) for part in key.split("/"))
    return f"{str(public_base_url).rstrip('/')}/{encoded_key}"


def get_content_type(file_path: Path, fallback: str = "application/octet-stream") -> str:
    guessed, _ = mimetypes.guess_type(str(file_path))
    return guessed or fallback


def upload_file_to_r2(
    file_path: str | Path,
    r2_config: dict[str, Any],
    filename: str | None = None,
    content_type: str | None = None,
) -> dict[str, Any]:
    path = Path(file_path)
    endpoint = normalize_endpoint(r2_config)
    if not endpoint:
        raise RuntimeError("Thieu endpoint hoac accountId trong cau hinh R2 runtime_configs")

    object_filename = filename or path.name
    key = build_r2_key(r2_config, object_filename)
    file_size = path.stat().st_size
    object_content_type = content_type or get_content_type(path)

    client = boto3.client(
        "s3",
        region_name=r2_config.get("region") or "auto",
        endpoint_url=endpoint,
        aws_access_key_id=r2_config.get("accessKeyId"),
        aws_secret_access_key=r2_config.get("secretAccessKey"),
        config=Config(signature_version="s3v4", s3={"addressing_style": "path"}),
    )

    part_size = int(r2_config.get("partSize") or SERVER_CONFIG.r2_upload_part_size)
    queue_size = int(r2_config.get("queueSize") or SERVER_CONFIG.r2_upload_queue_size)
    transfer_config = TransferConfig(
        multipart_threshold=part_size,
        multipart_chunksize=part_size,
        max_concurrency=queue_size,
        use_threads=True,
    )

    print(f"[R2] Uploading {path} -> {r2_config.get('bucket')}/{key}")
    client.upload_file(
        str(path),
        r2_config["bucket"],
        key,
        ExtraArgs={"ContentType": object_content_type},
        Config=transfer_config,
    )

    etag = None
    try:
        head = client.head_object(Bucket=r2_config["bucket"], Key=key)
        etag = head.get("ETag")
    except Exception:
        pass

    print(f"[R2] Upload complete: {r2_config.get('bucket')}/{key}")
    return {
        "bucket": r2_config["bucket"],
        "key": key,
        "size": file_size,
        "contentType": object_content_type,
        "etag": etag,
        "url": build_public_url(r2_config, key),
    }
